sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This library contains arithmetic for the finite field modulo the prime
+ * p = 2^255-19, which is used for both the X25519 key exchange and the Ed25519
+ * signature scheme.
+ */
+
+
+/**
+ * Fully reduce a 510-bit number modulo p.
+ *
+ * Returns c = a mod p.
+ *
+ * Uses Solinas/pseudo-Mersenne reduction, which is based on the observation
+ * that if a large number x is split into two so that the lower 255 bits are x0
+ * and all bits above 255 are x1, then:
+ *
+ *   x = x0 + (2^255 * x1) \equiv (x0 + (19 * x1)) (mod p)
+ *
+ * If x is large, then x0 + 19 * x1 will be much smaller than x, because 19 <<
+ * 2^255. This step can be repeated as necessary until a conditional
+ * subtraction is enough to fully reduce.
```

Comment:
I've been struggling to find a good reference for explaining Solinas reduction. But the way I think of it is as an iterative process, where the number of steps depends on the prime -- so a constant number of steps here, because the prime is fixed, but not constant across primes.

For a prime `2^n - K`, a step takes the bits above `2^n`, multiplies them by `K`, and adds them to the lower `n` bits. So if the current intermediate result is `b` bits, each step reduces the overall bound from `2^b` to `2^n + K * (2^(b-n))`. Here are some quick calculations in Python for this 25519 implementation (which first reduces by `2p=2^256-38` and then by `p=2^255-19`):
```python
>>> mask = lambda n: (1 << n) - 1
>>> def step_bound(prev_bound, n, k):
...   bits = int.bit_length(prev_bound)
...   return mask(n) + (k * mask(bits - n))
... 
>>> b0 = mask(510)               # starting bounds
>>> b1 = step_bound(b0, 256, 38) # bounds after first step using 2p
>>> int.bit_length(b1)
260
>>> b2 = step_bound(b1, 255, 19) # bounds after second step using p
>>> int.bit_length(b2)
256
>>> b2 < 2*((1 << 255) - 19)     # can we do a conditional subtraction now?
True
```
However, for e.g. P-256, where `n = 256` and `K` is a 224-bit number, you get less out of each step (compare with the 5-bit `K` in 25519, or even the `n = 384` and 128-bit `K` in P-384). For P-256, you actually need 8 steps:
```python
>>> K = (1<<224) - (1<<192) - (1<<96) + 1
>>> b0 = mask(256*2)
>>> b1 = step_bound(b0, 256, K)
>>> int.bit_length(b1)
480
>>> b2 = step_bound(b1, 256, K)
>>> int.bit_length(b2)
448
>>> b3 = step_bound(b2, 256, K)
>>> int.bit_length(b3)
416
>>> b4 = step_bound(b3, 256, K)
>>> int.bit_length(b4)
384
>>> b5 = step_bound(b4, 256, K)
>>> int.bit_length(b5)
352
>>> b6 = step_bound(b5, 256, K)
>>> int.bit_length(b6)
320
>>> b7 = step_bound(b6, 256, K)
>>> int.bit_length(b7)
288
>>> b8 = step_bound(b7, 256, K)
>>> int.bit_length(b8)
257
>>> b8 < 2*((1<<256) - K) # can we do a conditional subtraction now?
True
```
I think maybe, for want of a good link, I'll just put a big block comment at the top here explaining Solinas reduction in general. Or maybe a `solinas_reduction.md` in the `otbn/crypto` directory so that I can reference it both here and for P-384.